### PR TITLE
Autocomplete

### PR DESCRIPTION
**Pull request by Nathan-Fenner**
Autocomplete only knows about keywords right now (and some of the functions).

### CODE REVIEW

File: main/static/script.js [hunk 1]
```diff
@@ -60,6 +62,19 @@ module.controller("mainCtrl", function(
     renderType: "line"
   };
 
+  $scope.autocom = new Autocom(document.getElementById("queryInput"));
```
Comment (jeeyoungk) [inline main/static/script.js]: let's use dashes

File: main/static/script.js [hunk 2]
```diff
@@ -46,6 +46,8 @@ module.service("$google", function($rootScope) {
   }
 });
 
+var autocom;
```
Comment (jeeyoungk) [inline main/static/script.js]: this should go inside controller.

File: main/static/autocom.js [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+		} else {
+			tooltipState.active = false;
+		}
+		elements.tooltip.hidden = !tooltipState.active || tooltipSuppress;
+	}
+	var refresh = function() {
+		setTimeout(renderTooltip, 0);
+		var inputStyle = getComputedStyle(input);
+
+		// Make input have style that mtches the holder.
```
Comment (jeeyoungk) [inline main/static/autocom.js]: typo

File: main/static/autocom.js [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+
+		input.style.wordWrap = "normal";
+		elements.prior.style.wordWrap = input.style.wordWrap;
+		elements.after.style.wordWrap = input.style.wordWrap;
+
+		// Nowrap on the marker
+		elements.marker.style.whiteSpace = "nowrap";
+
+		// Input size
+		input.style.height = "100%";
```
Comment (jeeyoungk) [inline main/static/autocom.js]: we should probably leave `input` alone.

File: main/static/autocom.js [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+		input.style.overflowX = "hidden";
+
+		// Tooltip
+		elements.tooltip.style.position = "absolute";
+		elements.tooltip.style.zIndex = 1000;
```
Comment (jeeyoungk) [inline main/static/autocom.js]: zindex should be configurable (most annoying thing :P )

File: main/static/autocom.js [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+// scoreTable uses the table for memoization and computes a modified LCS on A and B.
+// The config contains parameters for this procedure.
+function scoreTable(table, A, B, i, j, config) {
+	if (i >= A.length || j >= B.length) {
+		return config.skipGiven * (i - A.length); // Don't penalize for the end of the second word
+	}
+	if (table[i][j] !== null) {
+		return table[i][j];
+	}
+	if (A[i] == B[j]) {
```
Comment (jeeyoungk) [inline main/static/autocom.js]: `===`

File: main/static/autocom.js [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+		prior:   prior,
+		marker:  marker,
+		after:   after,
+		tooltip: tooltip
+	};
+}
+
+// predictReady determines whether or not the input is ready for prediction,
+// and it asks for something that matches non-separators.
+function predictReady(input, letter) {
```
Comment (jeeyoungk) [inline main/static/autocom.js]: let's return `null` - feels weird to be returning `Either<Bool, Object>`

File: main/static/autocom.js [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+	var before = input.value.substring(0, at).match("(" + letter + ")+$");
+	if (before && before.length > 0) {
+		return {from: at - before[0].length, word: before[0], to: at};
+	}
+	return false;
+}
+
+// filterCandidates finds a list of candidates, pursuant to the config, among the options
+// from the given prefix `word`.
+function filterCandidates(word, options, config) {
```
Comment (jeeyoungk) [inline main/static/autocom.js]: rather return array all the time - much easier to deal with.

File: main/static/autocom.js [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+
+// Moves the tooltip to its location proper.
+function moveTooltip(elements, offsetX, offsetY) {
+	// Resize the shadower:
+	elements.shadow.style.width = getComputedStyle(elements.input).width;
+	// Fill prior and after with text:
+	elements.prior.innerHTML = elements.input.value.substring(0, elements.input.selectionStart);
+	elements.after.innerHTML = elements.input.value.substring(elements.input.selectionStart);
+	// Use marker's location to reposition tooltip:
+	elements.tooltip.style.left = ((elements.input.offsetLeft + elements.marker.offsetLeft + offsetX)|0) + "px";
```
Comment (jeeyoungk) [inline main/static/autocom.js]: should do `|| 0` instead

File: main/static/autocom.js [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+	elements.tooltip.style.top = ((elements.input.offsetTop + elements.marker.offsetTop - elements.input.scrollTop + offsetY)|0) + "px";
+}
+
+// Creates and adds to the tooltip one row, with the given word/index.
+// If clicked, it invokes the given selectedCallback.
+function generateTooltipRow(tooltip, word, index, selectedIndex, selectedCallback) {
+	var row = document.createElement("div");
+	row.className = "autocom-tooltip-line";
```
Comment (jeeyoungk) [inline main/static/autocom.js]: can we pull out the class names as CAPITALIZED variable names to the top?

File: main/static/autocom.js [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+			}
+			if (e.keyCode == 40) { // DOWN
+				e.preventDefault();
+				tooltipState.index++;
+				if (tooltipState.index >= tooltipState.words.length) {
+					tooltipState.index = 0;
+				}
+				return;
+			}
+			if (e.keyCode == 27) {
```
Comment (jeeyoungk) [inline main/static/autocom.js]: what's code 27?

Comment (jeeyoungk) [inline main/static/autocom.js]: oh it's ESC.

File: main/static/autocom.js (hunk 9 above)
Comment (Nathan-Fenner) [inline main/static/autocom.js]: `|0` is the same as `Math.floor`

Comment (jeeyoungk) [inline main/static/autocom.js]: ohhhh. nice.

**Comment by jeeyoungk** [comment]
LGTM

**Pull request merged**